File: scripts/mergesort/ms_seq.c
/* Mergesort sequential baseline in C — monolithic top-down recursive
 * merge sort on a flat int32 array. Below CUTOFF: insertion sort.
 * Above: split, recurse, merge into scratch, copy back.
 *
 * Build:  gcc -O3 -march=native -o ms_seq_c ms_seq.c
 * Run:    ./ms_seq_c DATA_DIR
 * Output: CHECKSUM=<sum> RUNTIME_SEC=<float>
 *
 * The checksum is sum-mod-2^32 of the array. Sum is order-invariant so it
 * cross-validates against expected_checksum.txt regardless of how sorted
 * the result is — but we ALSO verify monotonicity at the end to catch
 * bugs where the algorithm just shuffles or returns the input.
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>

static int N;
static int CUTOFF;

static int *arr;
static int *tmp;

static void load_config(const char *dir) {
    char path[1024]; snprintf(path, sizeof path, "%s/config.txt", dir);
    FILE *f = fopen(path, "r"); if (!f) { perror(path); exit(1); }
    char k[64]; int v;
    while (fscanf(f, "%63s %d", k, &v) == 2) {
        if      (!strcmp(k, "N"))      N      = v;
        else if (!strcmp(k, "CUTOFF")) CUTOFF = v;
    }
    fclose(f);
}

/* Insertion sort on arr[lo..hi). */
static inline void insertion_sort(int *a, int lo, int hi) {
    for (int i = lo + 1; i < hi; i++) {
        int x = a[i];
        int j = i - 1;
        while (j >= lo && a[j] > x) { a[j + 1] = a[j]; j--; }
        a[j + 1] = x;
    }
}

/* Merge a[lo..mid) and a[mid..hi) into t[lo..hi). */
static inline void merge_to(int *a, int lo, int mid, int hi, int *t) {
    int i = lo, j = mid, k = lo;
    while (i < mid && j < hi) {
        if (a[i] <= a[j]) t[k++] = a[i++];
        else              t[k++] = a[j++];
    }
    while (i < mid) t[k++] = a[i++];
    while (j < hi)  t[k++] = a[j++];
    memcpy(a + lo, t + lo, (size_t)(hi - lo) * sizeof(int));  /* copy back */
}

static void ms_sort(int *a, int lo, int hi) {
    if (hi - lo <= CUTOFF) { insertion_sort(a, lo, hi); return; }
    int mid = lo + (hi - lo) / 2;
    ms_sort(a, lo, mid);
    ms_sort(a, mid, hi);
    merge_to(a, lo, mid, hi, tmp);
}

int main(int argc, char **argv) {
    if (argc < 2) { fprintf(stderr, "usage: %s DATA_DIR\n", argv[0]); return 2; }
    const char *dir = argv[1];
    load_config(dir);

    arr = (int*)malloc((size_t)N * sizeof(int));
    tmp = (int*)malloc((size_t)N * sizeof(int));
    if (!arr || !tmp) { fprintf(stderr, "OOM\n"); return 1; }

    char inp[1024]; snprintf(inp, sizeof inp, "%s/input.bin", dir);
    FILE *f = fopen(inp, "rb"); if (!f) { perror(inp); return 1; }
    if (fread(arr, sizeof(int), N, f) != (size_t)N) {
        fprintf(stderr, "short read %s\n", inp); return 1;
    }
    fclose(f);

    /* Compute pre-sort checksum for sanity (must match post-sort). */
    uint64_t pre = 0;
    for (int i = 0; i < N; i++) pre = (pre + (uint64_t)(uint32_t)arr[i]) & 0xFFFFFFFFu;

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    ms_sort(arr, 0, N);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double secs = (double)(t1.tv_sec - t0.tv_sec) + (double)(t1.tv_nsec - t0.tv_nsec) / 1e9;

    /* Verify sorted and compute post-sort checksum. */
    uint64_t cs = 0;
    int ok = 1;
    for (int i = 0; i < N; i++) {
        if (i > 0 && arr[i] < arr[i - 1]) { ok = 0; }
        cs = (cs + (uint64_t)(uint32_t)arr[i]) & 0xFFFFFFFFu;
    }
    if (!ok) { fprintf(stderr, "FATAL: array not sorted\n"); return 1; }
    if (cs != pre) { fprintf(stderr, "FATAL: checksum changed (pre=%lu post=%lu)\n",
                              (unsigned long)pre, (unsigned long)cs); return 1; }

    printf("CHECKSUM=%lu\n", (unsigned long)cs);
    printf("RUNTIME_SEC=%f\n", secs);

    free(arr); free(tmp);
    return 0;
}

File: scripts/mergesort/ms_seq.hs
{-# LANGUAGE BangPatterns #-}
module Main where

-- Mergesort sequential baseline in Haskell — monolithic top-down
-- recursive merge sort on Data.Vector.Unboxed.Mutable Int32. Below
-- CUTOFF: insertion sort. Above: split, recurse, merge into scratch,
-- copy back. Same algorithm as ms_seq.c and ms_seq.rs.

import Control.Monad (forM_, when)
import Data.Bits ((.&.))
import Data.Int (Int32)
import Data.Word (Word32, Word64)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Internal as BSI
import qualified Data.Vector.Unboxed.Mutable as MV
import Foreign.ForeignPtr (withForeignPtr)
import Foreign.Ptr (Ptr, plusPtr)
import Foreign.Storable (peek)
import Data.Time.Clock (getCurrentTime, diffUTCTime)
import System.IO (hFlush, stdout)
import System.Environment (getArgs)
import System.IO (hPutStrLn, stderr)

readConfig :: FilePath -> IO (Int, Int)
readConfig path = do
  s <- readFile path
  let kv = [ (head ws, read (ws !! 1) :: Int)
           | l <- lines s, let ws = words l, length ws >= 2 ]
  let n      = maybe 0 id (lookup "N" kv)
      cutoff = maybe 64 id (lookup "CUTOFF" kv)
  return (n, cutoff)

readInput :: FilePath -> Int -> IO (MV.IOVector Int32)
readInput path n = do
  bs <- BS.readFile path
  let !(BSI.BS fp _) = bs
  v <- MV.replicate n 0
  withForeignPtr fp $ \src ->
    forM_ [0..n-1] $ \i -> do
      !w <- peek (src `plusPtr` (i*4) :: Ptr Int32)
      MV.write v i w
  return v

{-# INLINE insertionSort #-}
insertionSort :: MV.IOVector Int32 -> Int -> Int -> IO ()
insertionSort !a !lo !hi = go (lo + 1)
  where
    go !i
      | i >= hi   = return ()
      | otherwise = do
          !x <- MV.read a i
          let bubble !j
                | j <= lo   = MV.write a j x
                | otherwise = do
                    !y <- MV.read a (j - 1)
                    if y > x
                      then do MV.write a j y
                              bubble (j - 1)
                      else MV.write a j x
          bubble i
          go (i + 1)

{-# INLINE mergeTo #-}
mergeTo :: MV.IOVector Int32 -> Int -> Int -> Int -> MV.IOVector Int32 -> IO ()
mergeTo !a !lo !mid !hi !t = do
  let loop !i !j !k
        | i >= mid && j >= hi = return ()
        | i >= mid = do !y <- MV.read a j; MV.write t k y; loop i (j+1) (k+1)
        | j >= hi  = do !x <- MV.read a i; MV.write t k x; loop (i+1) j (k+1)
        | otherwise = do
            !x <- MV.read a i
            !y <- MV.read a j
            if x <= y then do MV.write t k x; loop (i+1) j     (k+1)
                      else do MV.write t k y; loop i     (j+1) (k+1)
  loop lo mid lo
  -- copy back t[lo..hi) into a[lo..hi)
  let cp !p
        | p >= hi   = return ()
        | otherwise = do !v <- MV.read t p; MV.write a p v; cp (p + 1)
  cp lo

msSort :: Int -> MV.IOVector Int32 -> Int -> Int -> MV.IOVector Int32 -> IO ()
msSort !cutoff !a !lo !hi !t
  | hi - lo <= cutoff = insertionSort a lo hi
  | otherwise = do
      let !mid = lo + (hi - lo) `div` 2
      msSort cutoff a lo  mid t
      msSort cutoff a mid hi  t
      mergeTo  a lo  mid hi  t

main :: IO ()
main = do
  args <- getArgs
  case args of
    (dir:_) -> run dir
    _       -> do hPutStrLn stderr "usage: ms_seq_hs DATA_DIR"; return ()

run :: FilePath -> IO ()
run dir = do
  (n, cutoff) <- readConfig (dir ++ "/config.txt")
  arr <- readInput (dir ++ "/input.bin") n
  tmp <- MV.replicate n 0

  -- Pre-sort checksum (sum mod 2^32)
  let presumGo !i !acc
        | i >= n    = return (acc :: Word64)
        | otherwise = do
            !x <- MV.read arr i
            let !w = fromIntegral (fromIntegral x :: Word32) :: Word64
            presumGo (i + 1) ((acc + w) .&. 0xFFFFFFFF)
  !pre <- presumGo 0 0

  t0 <- getCurrentTime
  msSort cutoff arr 0 n tmp
  t1 <- getCurrentTime
  let !secs = realToFrac (diffUTCTime t1 t0) :: Double

  -- Verify sorted + post-sort checksum
  let verifyGo !i !acc
        | i >= n    = return (acc :: Word64)
        | otherwise = do
            !x <- MV.read arr i
            when (i > 0) $ do
              !y <- MV.read arr (i - 1)
              when (x < y) $ do
                hPutStrLn stderr ("FATAL: not sorted at i=" ++ show i)
                error "not sorted"
            let !w = fromIntegral (fromIntegral x :: Word32) :: Word64
            verifyGo (i + 1) ((acc + w) .&. 0xFFFFFFFF)
  !cs <- verifyGo 0 0
  when (cs /= pre) $ do
    hPutStrLn stderr ("FATAL: checksum changed pre=" ++ show pre ++ " post=" ++ show cs)
    error "checksum mismatch"

  putStrLn ("CHECKSUM=" ++ show cs)
  putStrLn ("RUNTIME_SEC=" ++ show secs)
  hFlush stdout
